src/compute-client/src/controller/instance.rs

```diff
@@ -1117,15 +1117,65 @@ where
             self.compute
                 .send(ComputeCommand::CreateDataflow(augmented_dataflow));
 
-            // TODO: defer scheduling until inputs are ready
             for id in collections {
-                self.compute.send(ComputeCommand::Schedule(id));
+                self.maybe_schedule_collection(id);
             }
         }
 
         Ok(())
     }
 
+    /// Schedule the identified collection if all its inputs are available.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the identified collection does not exist.
+    fn maybe_schedule_collection(&mut self, id: GlobalId) {
+        let collection = self.compute.expect_collection(id);
+
+        // Don't schedule collections twice.
+        if collection.scheduled {
+            return;
+        }
+
+        let as_of = collection.read_frontier();
+
+        // If the collection has an empty `as_of`, it was either never installed on the replica or
+        // has since been dropped. In either case the replica does not expect any commands for it.
+        if as_of.is_empty() {
+            return;
+        }
+
+        let compute_frontiers = collection.compute_dependencies.iter().map(|id| {
+            let dep = &self.compute.expect_collection(*id);
+            &dep.write_frontier
+        });
+        let storage_frontiers = collection.storage_dependencies.iter().map(|id| {
+            let dep = &self.storage_controller.collection(*id).expect("must exist");
+            &dep.write_frontier
+        });
+        let ready = compute_frontiers
+            .chain(storage_frontiers)
+            .all(|frontier| PartialOrder::less_than(&as_of, &frontier.borrow()));
+
+        if ready {
+            self.compute.send(ComputeCommand::Schedule(id));
+            let collection = self.compute.expect_collection_mut(id);
+            collection.scheduled = true;
+        }
+    }
+
+    /// Schedule any unscheduled collections that are ready.
+    fn schedule_collections(&mut self) {
+        let collections = self.compute.collections.iter();
+        let candidates: Vec<_> = collections
+            .filter_map(|(id, c)| (!c.scheduled).then_some(*id))
+            .collect();
+        for id in candidates {
+            self.maybe_schedule_collection(id);
+        }
```

Comment:
The reason for having this prefilter is that it avoids doing a lookup in the `collections` map inside `maybe_schedule_collection` for collections that have already been scheduled (most of them most of the time). That avoids some work, but is maybe a premature optimization.

Happy to find a better way to structure this. Here are the constraints:
* We want a way to perform a targeted scheduling check for a given collection, so we can immediately schedule collections in `create_dataflow` if they are ready. This is was prompted `maybe_schedule_collection`.
* `maybe_schedule_collection` needs to take a `GlobalId` rather than a reference to a `CollectionState` to appease the borrow checker.

We can't have a `CollectionState::is_schedulable` because the scheduling check needs to look at the frontiers of dependencies, which `CollectionState` doesn't have access to. However, we could have an `Instance::is_collection_schedulable(&self, &CollectionState)` that's then called by both `create_dataflow` and `schedule_collections`. That would avoid the additional lookups and still be quite readable I think.

Reply:
Looks like this doesn't actually become more readable because of the additional code introduced at the two callsites. I simply removed the redundant filter now.